src/pudl/transform/eia861.py

```diff
@@ -2327,76 +2146,290 @@ def utility_data(tfr_dfs):
+        dfs=late_dfs,
+        cols=["report_date", "balancing_authority_code_eia", "utility_id_eia", "state"],
+    )
+    # We merge on ba_code then drop it, b/c only BA ID exists in all years consistently:
+    late_date_ba_util_state = (
+        late_date_ba_code_util_state.merge(late_ba_code_id, how="outer")
+        .drop("balancing_authority_code_eia", axis="columns")
+        .drop_duplicates()
+    )
+
+    ba_assn_eia861 = (
+        pd.concat([early_date_ba_util_state, late_date_ba_util_state])
+        .dropna(subset=["balancing_authority_id_eia"])
+        .pipe(apply_pudl_dtypes, group="eia")
+    )
+    return ba_assn_eia861
+
+
+@asset
+def balancing_authority_eia861(
+    clean_balancing_authority_eia861: pd.DataFrame,
+) -> pd.DataFrame:
+    """Finish the normalization of the balancing_authority_eia861 table.
+
+    The balancing_authority_assn_eia861 table depends on information that is only
+    available in the UN-normalized form of the balancing_authority_eia861 table, so and
+    also on having access to a bunch of transformed data tables, so it can compile the
+    observed combinations of report dates, balancing authorities, states, and utilities.
+    This means that we have to hold off on the final normalization of the
+    balancing_authority_eia861 table until the rest of the transform process is over.
+    """
+    logger.info("Completing normalization of balancing_authority_eia861.")
+    ba_eia861_normed = clean_balancing_authority_eia861.loc[
+        :,
+        [
+            "report_date",
+            "balancing_authority_id_eia",
+            "balancing_authority_code_eia",
+            "balancing_authority_name_eia",
+        ],
+    ].drop_duplicates(subset=["report_date", "balancing_authority_id_eia"])
+
+    # Make sure that there aren't any more BA IDs we can recover from later years:
+    ba_ids_missing_codes = (
+        ba_eia861_normed.loc[
+            ba_eia861_normed.balancing_authority_code_eia.isnull(),
+            "balancing_authority_id_eia",
+        ]
+        .drop_duplicates()
+        .dropna()
+    )
+    fillable_ba_codes = ba_eia861_normed[
+        (ba_eia861_normed.balancing_authority_id_eia.isin(ba_ids_missing_codes))
+        & (ba_eia861_normed.balancing_authority_code_eia.notnull())
+    ]
+    if len(fillable_ba_codes) != 0:
+        raise ValueError(
+            f"Found {len(fillable_ba_codes)} unfilled but fillable BA Codes!"
         )
-        tfr_dfs[tfr_func] = _early_transform(raw_dfs[tfr_func])
-        tfr_dfs = tfr_funcs[tfr_func](tfr_dfs)
-
-    # This is more like harvesting stuff, and should probably be relocated:
-    tfr_dfs = balancing_authority_assn(tfr_dfs)
-    tfr_dfs = utility_assn(tfr_dfs)
-    tfr_dfs = normalize_balancing_authority(tfr_dfs)
-    # Do some final cleanup and assign types.
-    return {
-        name: convert_cols_dtypes(df, data_source="eia") for name, df in tfr_dfs.items()
-    }
+
+    return _post_process(ba_eia861_normed)
+
+
```

Comment:
These asset factories ended up being too slow, given the work they were encapsulating, but is this basically how they're supposed to be constructed? Anything weird or bad in here?